integration_tests/conftest.py

```diff
@@ -74,3 +80,189 @@ def pipeline_columns(sim, populations) -> List[str]:
+        numerator: int,
+        bug_distribution: scipy.stats.rv_discrete,
+        no_bug_distribution: scipy.stats.rv_discrete,
+    ) -> float:
+        # We can be dealing with some _extremely_ unlikely events here, so we have to set numpy to not error
+        # if we generate a probability too small to be stored in a floating point number(!), which is known
+        # as "underflow"
+        with np.errstate(under="ignore"):
+            bug_marginal_likelihood = bug_distribution.pmf(numerator)
+            no_bug_marginal_likelihood = no_bug_distribution.pmf(numerator)
+
+        try:
+            return bug_marginal_likelihood / no_bug_marginal_likelihood
+        except (ZeroDivisionError, FloatingPointError):
+            return 1_000_000.0
+
+    @cache
+    def _fit_beta_distribution_to_ui(self, lb: float, ub: float) -> Tuple[float, float]:
+        assert lb > 0 and ub < 1
+        # Inspired by https://stats.stackexchange.com/a/112671/
+        def objective(x):
+            # np.exp ensures they are always positive
+            a, b = np.exp(x)
+            dist = scipy.stats.beta(a=a, b=b)
+
+            squared_error_lower = self._quantile_squared_error(dist, lb, 0.025)
+            squared_error_upper = self._quantile_squared_error(dist, ub, 0.975)
+
+            return squared_error_lower + squared_error_upper
+
+        # It is quite important to start with a reasonable guess.
+        ui_midpoint = (lb + ub) / 2
+        # TODO: Is this reasonable!?
+        for first_guess_concentration in [1_000, 100, 10]:
+            optimization_result = scipy.optimize.minimize(
+                objective,
+                x0=[
+                    np.log(ui_midpoint * first_guess_concentration),
+                    np.log((1 - ui_midpoint) * first_guess_concentration),
+                ],
+            )
+            # Sometimes it warns that it may not have found a good solution,
+            # but the solution is very accurate.
+            if optimization_result.success or optimization_result.fun < 1e-05:
+                break
+
+        assert optimization_result.success or optimization_result.fun < 1e-05
+
+        result = np.exp(optimization_result.x)
+        assert len(result) == 2
+        return tuple(result)
+
+    def _quantile_squared_error(
+        self, dist: scipy.stats.rv_continuous, value: float, intended_quantile: float
+    ) -> float:
+        with np.errstate(under="ignore"):
+            actual_quantile = dist.cdf(value)
+
+        if 0 < actual_quantile < 1:
+            return (
+                scipy.special.logit(actual_quantile) - scipy.special.logit(intended_quantile)
+            ) ** 2
+        else:
+            # In this case, we were so far off that the actual quantile can't even be
+            # precisely calculated.
+            # We return an arbitrarily large penalty to ensure this is never selected as the minimum.
+            return 1_000_000.0**2
+
+    def write_output(self) -> None:
```

Comment:
This is also just writing an output file to look at for further investigation to go along with the normal test suites having to pass correct?

Reply:
Yes, this is an output file a human could inspect, either just for visibility into how the fuzzy checking is working overall, or to see more details about the cause of warnings. I've proposed here that we output this file each time we run the integration tests, but do _not_ track it in source control, to avoid lots of diff/merge conflicts on what is essentially a build artifact.